include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -564,6 +568,111 @@ __pattern_minmax_element(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __c
+{
+    // The algorithm reduces values in [first2, first2 + (last1-first1)) where the
+    // associated keys for the values are equal to the adjacent key.
+    //
+    // Example: keys       = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 } -- [first1, last1)
+    //          values     = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 } -- [first2, first2+n)
+    //
+    //          out_keys   = { 1, 2, 3, 4, 1, 3, 1, 3, 0 } -- result1
+    //          out_values = { 1, 2, 3, 4, 2, 6, 2, 6, 0 } -- result2
+
+    if (__keys.empty())
+        return 0;
+
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+
+    // Round 1: reduce with extra indices added to avoid long segments
+    // TODO: Add a check of whether there are any long key subsequences, and skip a round of copy_if
+    // and reduces if there are none.
+    auto __n = __keys.size();
+    auto __idx = oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __diff_type>(__exec, __n)
+                     .get_buffer();
+    auto __tmp_out_keys =
+        oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __key_type>(__exec, __n).get_buffer();
+    auto __tmp_out_values =
+        oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n).get_buffer();
+
+    //create two views over keys, the first with the first key removed for adjacent key comparison
+    auto __k1 = oneapi::dpl::__ranges::drop_view_simple<_Range1, __diff_type>(__keys, 1);
+    auto __k2 = __keys;
+
+    // view1 elements are a tuple of the element index and pairs of adjacent keys
+    // view2 elements are a tuple of the elements where key-index pairs will be written by copy_if
+    auto __view1 = experimental::ranges::zip_view(experimental::ranges::views::iota(0, __n), __k1, __k2);
+    auto __view2 = experimental::ranges::zip_view(experimental::ranges::views::all_write(__tmp_out_keys),
+                                                  experimental::ranges::views::all_write(__idx));
+
+    // use workgroup size as the maximum segment size.
+    ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+    // change __wgroup_size according to local memory limit
+    __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+        ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+    // element is copied if it is the last element (marks end of final segment), is in an index
+    // evenly divisible by wg size (ensures segments are not long), or has a key not equal to the
+    // adjacenent element (marks end of real segments)
+    auto __res = __pattern_copy_if(::std::forward<_ExecutionPolicy>(__exec), __view1, __view2,
```

Comment:
Does `_res` contain the number of copied elements or the position of the last written element? It makes sense to use a descriptive name.

Reply:
`__res` is the number of elements copied, and thus the index one past the last element written. I'll replace it with `__result_end`.

Reply:
`__res` contains the number of copied elements, which is the position one past the last written element.  I'll change it to `__result_end`.

Reply:
Done.